### DISCUSSION TITLE
Is there a way to use the scratch addon locally for addon dev debugging?

### DISCUSSION THREAD

**Discussion by OriLiMu** [ScratchAddons/ScratchAddons#7541] [Help] [answered]
Since some network issue it takes a long time for me to load the project page. I want to make a addon now which means needs a lot of refresh the page in the process. 
So any way to run the scratch page locally? 

is `scratch-www` or `scratch-gui` can be used in this case? if the scratch addon can be added on it too? 

I found that in the `addon.json` file. 
The "Matches" for url may lay a strict on the url, so if access the scratch locally may need some change of the code. Is this a big work?

**Comment by Samq64** [comment, marked as answer]
I haven't tested myself but addons should run on localhost. The default port for scratch-gui has changed though. (See #7474)

**Reply by OriLiMu** [reply]
do you know Which one is easier to config `scratch-www` or `scratch-gui` or they are just the same except the port issue? 
And do you know the differences of these two ? Seems totally same to me.

**Reply by Samq64** [reply]
`scratch-gui` is the editor and `sratch-www` includes most of the website except for profiles and the forums. I wouldn't recommend setting up either locally unless you have to.

**Reply by Samq64** [reply]
If you need scratch-gui TurboWarp has a guide and is slightly easier to setup: https://docs.turbowarp.org/development/getting-started

**Reply by mxmou** [reply]
> I haven't tested myself but addons should run on localhost.

They currently don't (see #7455). This can be fixed by adding `http://localhost/*` to the `matches` array on line 77 of `manifest.json`.

`scratch-gui` is easier to set up locally than `scratch-www` but you will either need to change the port it runs on to 8601 (by setting the `PORT` environment variable or editing the webpack configuration) or replace all instances of 8601 in the Scratch Addons source code with 8602.

**Reply by OriLiMu** [reply]
> If you need scratch-gui TurboWarp has a guide and is slightly easier to setup: https://docs.turbowarp.org/development/getting-started
>
> 

You mean use turbowarp as the local dev environment . 

If I make a mod for TurboWarp, is it can be easily merged to the `ScratchAddon` which can be installed on the broswer?  Mainly I want to make some contribute to the Scratchaddon which can make me learn a lot from the great programmers.

**Reply by Samq64** [reply]
You don't need to mod TurboWarp you just need to setup `scratch-gui` to run on localhost. But if your internet is slow than that won't really help because `scratch-gui` requires downloading about gigabyte of dependencies.
